### DISCUSSION TITLE
Batching individual updates

### DISCUSSION THREAD

**Discussion by frankvollebregt** [simolus3/drift#1400] [Q&A] [answered]
Hello! I noticed my code running slowly and narrowed it down to me doing many small updates, the majority of which can probably be put into a batch. The problem is that I'm struggling to figure out a 'good' way to do it.

Essentially I have two tables, one referencing the other. Let's say there's `Apples` and `Trees`, where each `Apple` contains a reference to its `Tree`'s `TreeId`. Now `Tree` (for some backend sync reasons) also has a `TreeUuid`.

```sql
CREATE TABLE trees (
  treeId INT NULL,  -- filled in from backend
  treeUuid TEXT NOT NULL PRIMARY KEY -- not in backend, filled in, used as key in app
);

CREATE TABLE apples (
  ...,
  treeId INT NULL,  -- filled in from backend
  treeUuid TEXT NULL REFERENCES trees(treeUuid) -- not in backend, needs to be filled in by update
);
```

What the update should do, is update all `Apple`s, by finding the corresponding `Tree` and adding a reference to the `TreeUuid` to the `Apple`.

My current approach looks a bit like the following:
```dart
List<Tree> trees = await this.allTrees().get();

// for each tree
for (Tree tree in trees) {
  // update all apples for this particular tree
  update(apples)
    ..where((apple) => apple.treeId.equals(tree.treeId))
    ..write(ApplesCompanion(treeUuid: Value(tree.treeUuid)));
}
```

How can I go about doing this in a batch?

Should I use a custom query to join the apples and trees and use that to update? Since there's quite a few combinations of tables I need to process, that'd require quite a bit of manual query writing, which isn't ideal (but if it's the best solution then so be it).

Thanks in advance for the help :)

**Comment by North101** [comment, marked as answer]
This should work:
```
await batch((batch) {
      for (final tree in trees) {
        batch.update(
          apples,
          ApplesCompanion(treeUuid: Value(tree.treeUuid)),
          where: apples.treeId.equals(tree.treeId),
        );
      }
});
```

This would probably be even faster, though its SQL only (I don't think moor support `UPDATE FROM`):
```sql
UPDATE apples
SET
  tree_uuid = tree.tree_uuid
FROM trees
WHERE trees.tree_id = apples.tree_id
```

However, I think the better solution is to just remove `apples.tree_id` and instead get the `tree_id` from the `trees` table.

**Reply by simolus3** [reply]
Yes, this `batch` should work. There's no equivalent to `UPDATE FROM` in the Dart API, it should be supported in moor files though.

If you want to avoid rewriting things, I think you can get most of the performance benefits of a `batch` by just wrapping tiny updates in a `transaction` block. This doesn't require you to rewrite updates to use the batch api. An advantage of a `batch` is that statements aren't prepared multiple times. But typically, the atomic write necessary for each individual update is what's causing most of the costs, and you can eliminate that with a transaction alone.

**Reply by frankvollebregt** [reply]
Thank you both for the quick response and advice!

**Reply by frankvollebregt** [reply]
In order to increase the performance, I ended up writing the queries in my `.moor` files using the `UPDATE FROM` as suggested. I'm performing this operation for 23 different relations (some of which with the same 'tree' table, none depending on the result of previous ones).

These operations (together) still take about 1.0 to 1.4 seconds, which is faster than before but feels a lot slower than it should be. Each table has up to 50 records (most not even half that). Would you expect this, or is it an indication that I'm still doing something wrong?

In Dart, my code to execute the queries looks like this:
```dart
print('Setting UUID keys for all tables');
await Future.wait([
      this.linkUuidTreesApples(),
      this.linkUuidTreesPears(),
      ... // 21 more
]);
```

Where these functions are defined in my moor file like:
```sql
linkUuidTreesApples: UPDATE apples
SET
    tree_uuid= tree.tree_uuid
FROM trees
WHERE trees.tree_id = apples.tree_id;
```
The internet points me towards defining indices, but with the small data size I doubt this will make a significant difference in my case.

@simolus3 do you have an idea how I can improve the performance of these operations?

**Reply by davidmartos96** [reply]
Maybe the problem is the structure/squema of your database. It sounds strange that 1 update affects 23 relations.
For example, instead of an apples table and pears table you could have a single table with an additional column that specifies the fruit with an ID (1 for apples and 2 for pears)

**Reply by davidmartos96** [reply]
Also, have you tried using a transaction like Simon suggested?

```dart
await this.transaction(() async {
    await Future.wait(....)
});
```

**Reply by frankvollebregt** [reply]
Hi David. Thanks for your suggestion!

Essentially, throughout my data model (which I get from the server) the tables are linked using regular (auto-inc) ids, and in the app I want to use uuids instead, such that I can add entries without causing any conflicts. Right after inserting my data, I then execute these 23 queries to set the uuids where appropriate.

Unfortunately, using a transaction like you suggested (either with or without the `Future.wait` statement) does not affect the running time of the code for me.

**Reply by frankvollebregt** [reply]
And in actuality, `apples` and `pears` are just an abstraction for orders, order lines, activities, products and all sorts of tables used in my actual database schema, to make the question a bit clearer without any context or knowledge of the actual system

**Reply by frankvollebregt** [reply]
Well now I feel stupid.

After struggling with this issue for longer than I'd like to admit, I started commenting out part of the operations, ending up only updating `apples` (so 1 out of the 23 operations), followed by not updating any of them. The operation then still took 1.5 seconds. That can't be right, I said to myself.

It turns out I timed not only these update operations, but also the statement where I enabled foreign keys, which takes up all the time.

The operation (wrapped in a transaction, thanks @davidmartos96 ;)) takes about 0.004 seconds now...

I'm not sure why or if enabling foreign keys should take so long, but it's definitely a separate issue from this thread, so if I don't figure it out I'll start a new discussion thread.

**Reply by davidmartos96** [reply]
@frankvollebregt Nice find :) You could start the investigation by taking a look at the output of `EXPLAIN QUERY PLAN UPDATE ....`. Any `SCAN TABLE` you see without an index is essentially looking at each row one by one
